### PR TITLE
add livestreaming support

### PR DESCRIPTION
**Pull request by pussinboot**
hey, apologies for the delay had some trouble duplicating your nginx config,, and now i see you've updated docs on how everything is set up

this is probably way more than what's needed to "just add streaming" but hopefully it should enable something to happen in a more hands-off way

main addition is entire user (dj) management system, only admin can create accounts and only dj's have access to the endpoint to start and stop streaming and only if no one else is streaming

they can also fill in some info about their stream/themselves and have it updated to the stream
maybe all you really need is the playlist_for & schedule changes but then you'll have to run a script on the server any time someone wants to stream ¯\_(ツ)_/¯

### CODE REVIEW

**Comment by barrucadu** [comment]
Thanks for this, this looks really cool.  I might not get chance to look at it today, but definitely tomorrow I'll do some code review.

File: frontend/js/radio.js [hunk 1]
```diff
@@ -156,6 +188,62 @@ function change_channel(e) {
...
+            sched_cell.style.width = "75%";
+
+            tbody.appendChild(row);
+        }
+
+        tbl.appendChild(tbody);
+        schedule_div.appendChild(tbl);
+
+        let explanation = document.createElement("h2");
+        explanation.innerHTML = "<br>All times are in UTC";
```
Comment (barrucadu) [inline frontend/js/radio.js]: `<h2><br>...</h2>` feels like a fairly hacky way to achieve whatever effect it is you want.  Is this really the best way?  Or would some CSS be better?

File: scripts/backend.py [hunk 2]
```diff
@@ -14,10 +14,23 @@
 """
 
 from docopt import docopt
-from flask import Flask, make_response, redirect, request, send_file
+from flask import Flask, make_response, redirect, request, send_file, render_template
 from mpd import MPDClient
 import json, os, random, time
 
+### all of these are added dependencies by puss, stdlib
+# import threading  # uncomment this if you want stream to start after current song ends
```
Comment (barrucadu) [inline scripts/backend.py]: I think the "start after current song ends" code can be removed, as in practice people will never start their stream at precisely the right second.

File: scripts/backend.py [hunk 3]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+    resp = make_response(json.dumps(sched_info), 200)
+    resp.headers["Content-Type"] = "application/json"
+    return resp
+
+
+def start_streaming():
+    # right now, this immediately stop the stream if you call actually_start_streaming instead
+    # this can wait for current track to finish playing if you uncomment below instead 
+    actually_start_streaming()
+    return
```
Comment (barrucadu) [inline scripts/backend.py]: Similarly, this code can be removed as people will start streaming whenever they do, not necessarily when a song ends.

File: scripts/backend.py [hunk 4]
```diff
@@ -28,6 +41,42 @@
...
+    exit(1)
+
+
+def handle_imminent_shutdown(*args):
+    pickle_path = os.path.join(db.SAVEDATA_PATH, 'livestream_save.pickle')
+    os.remove(pickle_path)  # something went terribly wrong so better to get rid of our savedata
+    exit(2)
+
+
+signal.signal(signal.SIGRTMIN, handle_imminent_shutdown)
```
Comment (barrucadu) [inline scripts/backend.py]: I've never caught `SIGRTMIN` before, is there any point in doing so?  How common is it?

File: scripts/backend.py [hunk 5]
```diff
@@ -49,23 +98,43 @@ def random_file_from(dname, cont=None):
     return cont(fname)
 
 
-def playlist_for(channel, beforeNum=5, afterNum=5):
-    """Return the playlist of the given MPD instance, as JSON."""
-
-    client = channels[channel]["client"]
+def get_playlist_info(client, beforeNum, afterNum, current_only=False):
```
Comment (barrucadu) [inline scripts/backend.py]: The `current_only` parameter and logic can be removed given my other comments.

File: scripts/backend.py [hunk 6]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
     return send_file(in_http_dir("404.html"))
 
 
+# streaming stuff begins
+# streaming overrides cyberia? (configure this in DEFAULT_LIVESTREAM_INFO)
+
+@app.route("/schedule.json", methods=["GET"])
+def schedule():
```
Comment (barrucadu) [inline scripts/backend.py]: This method is pretty dense, could you add a couple of comments?

File: scripts/backend.py [hunk 7]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+    # except:
+    #     print('couldn\'t connect to mpd')
+    #     return
+
+    # pinfo = get_playlist_info(client, 0, 0, True)
+    # wait_for_how_long = float(pinfo['current']['time']) - float(pinfo['elapsed'])
+    # threading.Timer(wait_for_how_long, actually_start_streaming).start()
+
+
+def actually_start_streaming():
```
Comment (barrucadu) [inline scripts/backend.py]: Can be inlined into `start_streaming`, given other comments.

File: scripts/backend.py [hunk 8]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+    # need to remake the client otherwise
+    # mpd.ConnectionError: Connection lost while reading line
+    try:
+        client = MPDClient()
+        client.connect(args["--mpd-host"], LIVESTREAM_INFO['PORT'])
+    except:
+        print('couldn\'t connect to mpd')
+        return
+
+    client.pause(1)
```
Comment (barrucadu) [inline scripts/backend.py]: I've had some more thoughts about this, and I guess MPD doesn't actually need to be paused and its outputs disabled.  It can just keep playing while the stream is happening (people will be able to access it at `/channel-cyberia.mp3` or something if they want).

Icecast can handle switching between the two audio sources automatically, which then removes the need for this logic.

File: scripts/backend.py [hunk 9]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+@login_required
+def streaming_page():
+    if LIVESTREAM_INFO['current_dj'] is None:
+        LIVESTREAM_INFO['current_dj'] = current_user.id
+        start_streaming()
+        return_text = "stream started, you can connect to icecast now"
+        # return_text = "you should be able to connect to icecast as soon as the currently playing track ends"
+    else:
+        return_text = 'someone else is already streaming..'
+    return return_text
```
Comment (barrucadu) [inline scripts/backend.py]: Rephrase this as

```python
if ...:
    ...
    return "..."

return "..."
```

File: scripts/backend.py [hunk 10]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+
+
+# maybe this should happen automatically if playlist update doesn't work for 30 seconds?
+@app.route("/dj/stop_streaming")
+@login_required
+def streaming_over_page():
+    if (LIVESTREAM_INFO['current_dj'] == current_user.id) or (current_user.id == 'admin'):
```
Comment (barrucadu) [inline scripts/backend.py]: Don't need the parentheses

File: scripts/backend.py [hunk 11]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+        stop_streaming()
+        return_text = 'stream ended'
+    else:
+        return_text = 'nop'
+    return return_text
```
Comment (barrucadu) [inline scripts/backend.py]: Reorganise as in `streaming_page`.

File: scripts/backend.py [hunk 12]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+
+
+def get_livestream_info():
+    # get current track info
+    req_url = "http://54.208.6.108:8000/status-json.xsl"  # temp dev
+    # req_url = "/radio/status-json.xsl"
```
Comment (barrucadu) [inline scripts/backend.py]: Your IP should probably be gone.

File: scripts/backend.py [hunk 13]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+            if 'stream_title' in dj_info and len(dj_info['stream_title']) > 0:
+                status_metadata['album'] = dj_info['stream_title']
+
+    if 'source' not in req_info['icestats']:
+        # nothing streaming yet
+        return {'current': status_metadata, 'elapsed': 0, 'stream_data': stream_data}
+
+    sources = req_info['icestats']['source']
+    for source in sources:
+        if 'cyberia' in source['listenurl']:
```
Comment (barrucadu) [inline scripts/backend.py]: Either have "cyberia" be a parameter to this function, or have a constant at the top-level.  Implicitly treating cyberia as the channel for streaming could lead to confusion later.

File: scripts/backend.py [hunk 14]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+            for k in status_metadata:
+                if k in source:
+                    status_metadata[k] = source[k].strip()
+            break
+
+    # if no last track or it's not the same as the last track append
+    if (len(LIVESTREAM_INFO['last_played']) == 0) or (LIVESTREAM_INFO['last_played'][-1].title != status_metadata['title']):
```
Comment (barrucadu) [inline scripts/backend.py]: Don't need the parentheses.

File: scripts/backend.py [hunk 15]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+        newest_track = LivestreamTrack(status_metadata['artist'], status_metadata['title'], time.time())
+        LIVESTREAM_INFO['last_played'].append(newest_track)
+        # we only want 5 last tracks + current
+        if (len(LIVESTREAM_INFO['last_played']) > 6):
```
Comment (barrucadu) [inline scripts/backend.py]: Parentheses.

File: scripts/backend.py [hunk 16]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+            LIVESTREAM_INFO['last_played'].pop(0)
+
+    time_now = time.time()
+
+    pinfo = {
+        'current': status_metadata,
+        'elapsed': int(time_now - LIVESTREAM_INFO['last_played'][-1].first_seen),
+        'before': [],
+        # "after":
```
Comment (barrucadu) [inline scripts/backend.py]: Remove this comment.

File: scripts/backend.py [hunk 17]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+    back_to_login = send_file(in_http_dir("login.html"))
+    if request.method == 'GET':
+        return back_to_login
+
+    username = request.form['username']
+    password = request.form['password']
+
+    check_user = db.DJUser.get(username)
+    if (check_user is not None):
+        if (check_user.password == password):
```
Comment (barrucadu) [inline scripts/backend.py]: Parentheses

File: scripts/backend.py [hunk 18]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+@login_required
+def logout():
+    logout_user()
+    return redirect("/")
+
+
+@app.route("/admin", methods=['GET', 'POST'])
+@login_required
+def admin_page():
+    if (current_user.id != 'admin'):
```
Comment (barrucadu) [inline scripts/backend.py]: Parentheses

File: scripts/backend.py [hunk 19]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+            return 'something went wrong'
+        return render_template("dj_page_edit.html", livestream_info=LIVESTREAM_INFO,
+                               dj_info_dict=dj_info, dj_info_names=djin,
+                               current_desc=dj_info['stream_desc'])
+    else:
+        did_it_work = db.update_dj_info(current_user.id, request.form.to_dict())
+        if did_it_work:
+            return redirect('/dj')
+        else:
+            return 'whoops, something went wrong'
```
Comment (barrucadu) [inline scripts/backend.py]: Is there a more meaningful error which `update_dj_info` could produce here?

File: scripts/backend.py [hunk 20]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+                db.change_password(current_user.id, new_pass)
+                logout_user()
+                return 'successfully changed your password, you\'ll need to log back in'
+        return redirect('/dj/password_change_form')
+
+
+@app.route("/admin/ban/<username>")
+@login_required
+def ban_user(username=None):
+    if (current_user.id != 'admin'):
```
Comment (barrucadu) [inline scripts/backend.py]: Parentheses

Comment (barrucadu) [inline scripts/backend.py]: Also, rather than only have one admin user with a fixed username, could users have an admin flag?

File: scripts/backend.py [hunk 21]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+        return redirect("/")
+    if username is not None:
+        ban_result = db.ban_user(username, True)
+        if ban_result is not None:
+            return '{} is now banned'.format(username)
+    return 'nothing happened'
```
Comment (barrucadu) [inline scripts/backend.py]: Probably better to give a more meaningful error message about the user not being found.

File: scripts/backend.py [hunk 22]
```diff
@@ -169,6 +238,312 @@ def page_not_found(error):
...
+@app.route("/admin/unban/<username>")
+@login_required
+def unban_user(username=None):
+    if (current_user.id != 'admin'):
+        return redirect("/")
+    if username is not None:
+        ban_result = db.ban_user(username, False)
+        if ban_result is not None:
+            return '{} is now unbanned'.format(username)
+    return 'nothing happened'
```
Comment (barrucadu) [inline scripts/backend.py]: Probably better to give a more meaningful error message about the user not being found.

File: scripts/templates/styles.css [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+    width: 100%;
+}
+
+#dj_info_edit > p > * {
+    width: 40%;
+}
+
+#stream_desc {
+    height: 75px;
+}
```
Comment (barrucadu) [inline scripts/templates/styles.css]: End this file with a newline.

File: scripts/database.py [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+# for url checking
+# slightly modified from https://stackoverflow.com/a/7160778
+url_regex = re.compile(r'^(?:http)s?://'  # http:// or https://
+                       r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # domain...
+                       r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'  # ...or ip
+                       r'(?::\d+)?'  # optional port
+                       r'(?:/?|[/?]\S+)$', re.IGNORECASE)
+
+
+def validate_url(url):
```
Comment (barrucadu) [inline scripts/database.py]: I'm not sure I see the point of checking a URL for well-formedness at all.

It's like checking emails for well-formedness. It doesn't matter if someone entered a valid email address, it matters if they get messages which are sent to it!

Similarly, it doesn't matter if someone enters a well-formed URL for their DJ picture, it matters if they entered the right URL. Well-formedness doesn't help you there.

File: scripts/database.py [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+    if len(url) == 0:
+        return True
+    return url_regex.match(url) is not None
+
+
+def get_ban_list():
+    all_user_info = THE_DB.all()
+    tor = []
+    for user in all_user_info:
+        if user['id'] != 'admin':
```
Comment (barrucadu) [inline scripts/database.py]: See earlier comment about admin users.

File: scripts/database.py [hunk 26]
```diff
@@ -0,0 +1,182 @@
...
+                banned = user['banned']
+            tor.append((user['id'], banned))
+
+    return tor
+
+
+def make_user(username):
+    check_query = Query()
+    check_if_user_exists = THE_DB.search(check_query.id == username)
+    if (len(check_if_user_exists) > 0):
```
Comment (barrucadu) [inline scripts/database.py]: Parentheses.

File: scripts/database.py [hunk 27]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    THE_DB.insert(new_user)
+    return (username, password)
+
+
+def get_dj_info(username):
+    check_query = Query()
+    check_if_user_exists = THE_DB.search(check_query.id == username)
+    if (len(check_if_user_exists) == 0):
```
Comment (barrucadu) [inline scripts/database.py]: Parentheses.

File: scripts/database.py [hunk 28]
```diff
@@ -0,0 +1,182 @@
...
+        return
+    return check_if_user_exists[0]
+
+
+def update_dj_info(username, form_dict):
+    check_query = Query()
+    check_if_user_exists = THE_DB.search(check_query.id == username)
+    if (len(check_if_user_exists) == 0):
```
Comment (barrucadu) [inline scripts/database.py]: Parentheses.

File: scripts/database.py [hunk 29]
```diff
@@ -0,0 +1,182 @@
...
+        if not validate_url(form_dict['dj_pic']):
+            del form_dict['dj_pic']
+    THE_DB.update(form_dict, check_query.id == username)
+    return True
+
+
+def change_password(username, new_pass=None):
+    check_query = Query()
+    check_if_user_exists = THE_DB.search(check_query.id == username)
+    if (len(check_if_user_exists) == 0):
```
Comment (barrucadu) [inline scripts/database.py]: Parentheses.

File: scripts/database.py [hunk 30]
```diff
@@ -0,0 +1,182 @@
...
+    if new_pass is None:
+        new_pass = "%032x" % random.getrandbits(128)
+    THE_DB.update({'password': new_pass}, check_query.id == username)
+    return new_pass
+
+
+def ban_user(username, new_ban_status):
+    check_query = Query()
+    check_if_user_exists = THE_DB.search(check_query.id == username)
+    if (len(check_if_user_exists) == 0):
```
Comment (barrucadu) [inline scripts/database.py]: Parentheses.

File: scripts/database.py [hunk 31]
```diff
@@ -0,0 +1,182 @@
...
+    return True
+
+
+def change_password(username, new_pass=None):
+    check_query = Query()
+    check_if_user_exists = THE_DB.search(check_query.id == username)
+    if (len(check_if_user_exists) == 0):
+        return
+    if new_pass is None:
+        new_pass = "%032x" % random.getrandbits(128)
```
Comment (barrucadu) [inline scripts/database.py]: What's this `%032x`?

File: scripts/database.py [hunk 32]
```diff
@@ -0,0 +1,182 @@
...
+            found_user = DJUser(db_res[0]['id'], db_res[0]['password'])
+        else:
+            found_user = None
+        return found_user
+
+
+if __name__ == '__main__':
+    new_admin = make_user('admin')
+    if new_admin is not None:
+        print('your admin account is now setup')
```
Comment (barrucadu) [inline scripts/database.py]: "set up" (or "set-up") is two words ("setup" is a different, single, word).

File: scripts/database.py [hunk 33]
```diff
@@ -0,0 +1,182 @@
...
+        self.password = password
+
+    @property
+    def is_active(self):
+        # check ban status
+        check_ban = Query()
+        check_ban_res = THE_DB.search(check_ban.id == self.id)
+        if len(check_ban_res) == 0:
+            return False
+        if 'banned' not in check_ban_res[0]:
```
Comment (barrucadu) [inline scripts/database.py]: I feel this should be a boolean, rather than a substring comparison.  A user could just have an `is_active` flag, rather than this more general notion of banning.

File: scripts/database.py [hunk 34]
```diff
@@ -0,0 +1,182 @@
...
+            return True
+        return not check_ban_res[0]['banned']
+
+    @property
+    def is_authenticated(self):
+        return True
+
+    @property
+    def is_anonymous(self):
```
Comment (barrucadu) [inline scripts/database.py]: Is this needed?  What would an anonymous user be?

File: scripts/database.py [hunk 35]
```diff
@@ -0,0 +1,182 @@
...
+        check_ban = Query()
+        check_ban_res = THE_DB.search(check_ban.id == self.id)
+        if len(check_ban_res) == 0:
+            return False
+        if 'banned' not in check_ban_res[0]:
+            return True
+        return not check_ban_res[0]['banned']
+
+    @property
+    def is_authenticated(self):
```
Comment (barrucadu) [inline scripts/database.py]: Is this needed?  Could it ever be `False`?

File: scripts/database.py [hunk 36]
```diff
@@ -0,0 +1,182 @@
...
+
+    def __eq__(self, other):
+        '''
+        Checks the equality of two `DJUser` objects using `get_id`.
+        '''
+        if isinstance(other, DJUser):
+            return self.get_id() == other.get_id()
+        return False
+
+    def __ne__(self, other):
```
Comment (barrucadu) [inline scripts/database.py]: Doesn't the default `__ne__` suffice here?

File: scripts/templates/admin.html [hunk 37]
```diff
@@ -0,0 +1,41 @@
...
+        <th>user</th>
+        <th>(❌ = b&amp;)</th> 
+      </tr>
+        {% for username, ban_status in all_users %}
+            <tr>
+               <th> {{ username }} </th>
+                {% if ban_status %}
+                   <td> <a href="/admin/unban/{{ username }}">❌</a></td>
+                {% else %}
+                   <td> <a href="/admin/ban/{{ username }}">🔨</a></td>
```
Comment (barrucadu) [inline scripts/templates/admin.html]: I think these icons are the wrong way around.  ❌ looks negative, so I'd read it as "disable this account".  Just using words is probably clearest here.  "disable" and "enable" would do fine.

File: scripts/templates/admin.html [hunk 38]
```diff
@@ -0,0 +1,41 @@
...
+             placeholder="username" required>
+         </div>
+     </div>
+    
+     <div class="control-group">
+         <div class="controls">
+           <button type="submit" class="btn btn-success">create</button>
+         </div>
+     </div>
+   </form>
```
Comment (barrucadu) [inline scripts/templates/admin.html]: In the frontend I use `form/fieldset/ol/li` to group controls.  It would be good to do the same in the backend for consistency.

File: scripts/templates/change_password.html [hunk 39]
```diff
@@ -0,0 +1,27 @@
...
+
+<head>
+  <title>change your password</title>
+</head>
+
+<body>
+<form action="" method=post class="form-horizontal">
+   <div class="control-group">
+       <div class="controls">
+         <p>current password <input type="password" name="current_pass" required></p>
```
Comment (barrucadu) [inline scripts/templates/change_password.html]: Use `label`s here, so that clicking the label selects the `input`.

**Comment by barrucadu** [comment]
**Note for me:** before deploying this, set up explicit mount point configuration with the channels at `/$channel-channel.{mp3,ogg}` as a fallback source: http://icecast.org/docs/icecast-2.4.1/config-file.html#mountsettings

This will allow streamers to just start streaming to the source and Icecast will handle the switchover, without needing to do anything to MPD.

**Comment by pussinboot** [comment]
Alright, I'll work on those changes.

The thing I'm most concerned about is whether the `get_livestream_info` function will work as intended because it's getting called any time someone requests the playlist json, would it be better to have a separate thread that gets created once a stream is started to scrape every so often from the icecast status.json? I think the way I have it now is defensive enough and tested with both my laptop and desktop connecting at different times but it may pose an issue if a lot of users are on the site, I'll have to redo the start and stop streaming functions either way with the fallback channel configuration change.

File: scripts/backend.py (hunk 4 above)
Comment (pussinboot) [inline scripts/backend.py]: I've never had it happen before, will remove `handle_imminent_shutdown` then

File: scripts/backend.py (hunk 12 above)
Comment (pussinboot) [inline scripts/backend.py]: whoops, dw it's my vps

File: scripts/backend.py (hunk 13 above)
Comment (pussinboot) [inline scripts/backend.py]: I'll add a parameter for the exact mount-name and it'll be an explicit equality check to avoid mishaps with the cyberia-channel.ogg fallback

File: scripts/database.py (hunk 24 above)
Comment (pussinboot) [inline scripts/database.py]: since whatever gets inputted here gets thrown inside the img source tag for the dj picture element i didn't want broken images but you're right that this doesn't guarantee that

File: scripts/database.py (hunk 31 above)
Comment (pussinboot) [inline scripts/database.py]: convert random 128 bits to a 32 hex digit password

File: scripts/database.py (hunk 33 above)
Comment (pussinboot) [inline scripts/database.py]: `check_ban_res[0]` is a dictionary, so first I'm checking that the key exists

File: scripts/database.py (hunk 35 above)
Comment (pussinboot) [inline scripts/database.py]: https://flask-login.readthedocs.io/en/latest/_modules/flask_login/mixins.html#UserMixin

these are all the functions required by flask_login

File: scripts/backend.py (hunk 20 above)
Comment (pussinboot) [inline scripts/backend.py]: sure, good idea

**Comment by barrucadu** [comment]
That's a good point about `get_livestream_info`, and probably worth doing.  A second thread could be created to update the MPD info every 1s and the stream info (if relevant) every 5s, say?

Also, I've made [the fallback channel change](https://github.com/barrucadu/nixfiles/commit/cb8c9098402471b00a28deaf883ed0c52e6b5dd7).  So now (eg) the MPD instance for cyberia streams to `/radio/mpd-cyberia.mp3`, and `/radio/cyberia.mp3` treats that as a fallback source if nobody is streaming directly to it.  Listeners are still subscribed as before to `/radio/cyberia.mp3`, so changing between DJ streaming and MPD streaming should be pretty seamless.

**Comment by pussinboot** [comment]
no need to drop the ogg streams : (
in my testing i used edcast which supported multiple uploads at once and audio isn't that much bandwidth

**Comment by barrucadu** [comment]
Ok, they're back.

**Comment by pussinboot** [comment]
hey, question before i do something unnecessary because of a potential misconfiguration

i set up icecast mounts exactly like you have it here https://github.com/barrucadu/nixfiles/blob/master/hosts/lainonlife/radio.nix 
if i connect to /radio/cyberia.ogg when i'm not streaming it falls back to mpd-cyberia.ogg as it should, but when i start streaming it doesn't automatically connect and i have to refresh
on the other hand if i'm connected to /radio/cyberia.ogg while streaming and disconnect, it crossfades back to whatever is playing on mpd-cyberia, but then reconnecting to the stream crossfades to the stream audio

i'm pretty sure the 2nd behavior is how it should work in the 1st case as well since `fallback-override` is set to 1, but if it's not i can have the player reconnect as soon as status.json notices a stream is live. that's kinda hacky though so i'd rather not lol.

**Comment by pussinboot** [comment]
what's weird is that if i open up the direct link to /radio/cyberia.ogg or what have you it works fine, the previously described 1st behavior only happens if i'm using the html audio player

**Comment by barrucadu** [comment]
There's now a conflict because I removed the hard-coded channels list in aa6cebf, but that should be an easy fix.

File: scripts/backend.py [hunk 40]
```diff
@@ -46,25 +93,151 @@ def random_file_from(dname, cont=None):
...
+    id='playlist_update',
+    name='Update [channel].json\'s',
+    replace_existing=True)
+
+
+def playlist_for(channel):
+    """Return the playlist of the given channel, as JSON."""
+    global channels
+
+    pinfo = channels[channel]["cache"]
```
Comment (barrucadu) [inline scripts/backend.py]: There's a race condition here, where someone requests the playlist before `update_mpd_info` has happened (even though it's only a second).  You could use `channels[channel].get("cache")`, which defaults to `None` if there is no value.

File: frontend/js/radio.js [hunk 41]
```diff
@@ -54,11 +54,13 @@ function check_status() {
             // Assume that the listeners of the ogg and mp3 streams
             // are disjoint and just add them.  Bigger numbers are
             // better, right?
-            let sname = source.server_name.substr(0, source.server_name - 6);
-            if(sname == channel || sname == "[mpd] " + channel) {
-                listeners     += source.listeners;
-                listenersPeak += source.listener_peak;
-                description    = source.server_description;
+            if (source.server_name !== undefined){
+                let sname = source.server_name.substr(0, source.server_name - 6);
```
Comment (barrucadu) [inline frontend/js/radio.js]: This is still a little broken, it should be `source.server_name.length - 6`
